pkg/infra/kvstore/kvstore.go

```diff
@@ -0,0 +1,65 @@
+package kvstore
+
+import (
+	"context"
+
+	"github.com/grafana/grafana/pkg/registry"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+)
+
+func init() {
+	registry.RegisterService(&KVStore{})
+}
+
+// Key/value store service
+type KVStore struct {
```

Comment:
Hard to put my finger on exactly why with my relatively minimal context, but something is tugging at me, saying "this ought to be an interface." Maybe because my first guess is that that clients ought not have to care about the difference between `KVStore` and `NamespacedKVStore`.

Reply:
Or maybe `KVStoreBackend` _is_ the abstracted interface clients are supposed to depend on. But in that case, why do we have these struct types wrapping it?

I'm just confused about this collection of interrelated types - i feel like i'm missing something about the intended use pattern, as what's here seems like more than what's minimally necessary.

Reply:
The intention here is for the consumer to interact only with `KVStore`, and the backend potentially be switched out depending on configuration.  So eg we could have an alternative backend based on redis.  With the upcoming switch to Wire we may be able to handle that in ProvideService and just have the backends implement an interface instead.

Reply:
> The intention here is for the consumer to interact only with `KVStore`

But not `NamespacedKVStore`? My naive assumption would be that consumers would not know or care whether their k/v CRUD is namespaced, instead having that be a property of configuration mandated e.g. by dependency injection.

> With the upcoming switch to Wire we may be able to handle that in ProvideService and just have the backends implement an interface instead.

This really seems like what i was craving, both for this and the `Init()` comment below: a way of pushing this kind of initialization logic out to a choice made by the caller.

Reply:
> But not `NamespacedKVStore`? My naive assumption would be that consumers would not know or care whether their k/v CRUD is namespaced, instead having that be a property of configuration mandated e.g. by dependency injection

There are 2 ways I was thinking about this:

1. Make it easier for something consuming the service to define a namespace in one place, eg:

```
	kvStore := uss.KVStore.WithNamespace(0, "my_namespace")
```

Then interact with `kvStore` to get/set keys rather than having to pass org id and namespace multiple times.

2. Provide a namespaced store object to components like plugins that should be able to get/set keys within their namespace but not outside of it.

> > With the upcoming switch to Wire we may be able to handle that in ProvideService and just have the backends implement an interface instead.
> 
> This really seems like what i was craving, both for this and the `Init()` comment below: a way of pushing this kind of initialization logic out to a choice made by the caller.

I'll take a look at how this could work with Wire and whether that gives us a pathway to simplify things.

Reply:
> Provide a namespaced store object to components like plugins that should be able to get/set keys within their namespace but not outside of it.

Righto, this is the case i was naively imagining, with the main question being whether plugins ought to be aware that its k/v operations are namespaced. My intuition says no because that implies the existence of a clear decision rule for classifying a component as namespaceable-or-not, and it seems unlikely that such a rule exists.

> I'll take a look at how this could work with Wire and whether that gives us a pathway to simplify things.

Thank you! (@marefr you may have a take here)